### DISCUSSION TITLE
Should the `passive_deletes` always be set to `True` when using the write-only loader?

### DISCUSSION THREAD

**Discussion by greyli** [sqlalchemy/sqlalchemy#11581] [Usage Questions] [answered]
## Environment

- SQLAlchemy 2.0.27
- SQLite 3
- Python 3.12

## Describe the use case

With a one-to-many relationship, I create the following `User` and `Photo` models and use write only loader for the `user.posts` relationship:

```python
class User(Base):
    __tablename__ = 'user'

    id: Mapped[int] = mapped_column(primary_key=True)
    username: Mapped[str] = mapped_column(String(20), unique=True, index=True)
    email: Mapped[str] = mapped_column(String(254), unique=True, index=True)

    photos: WriteOnlyMapped['Photo'] = relationship(back_populates='author', cascade='all, delete-orphan')


class Photo(Base):
    __tablename__ = 'photo'

    id: Mapped[int] = mapped_column(primary_key=True)
    description: Mapped[Optional[str]] = mapped_column(String(500))

    author_id: Mapped[int] = mapped_column(ForeignKey('user.id'))

    author: Mapped['User'] = relationship(back_populates='photos')
```

Now when I try to delete a user object, it raises the following exception ([related source code](https://github.com/sqlalchemy/sqlalchemy/blob/ee86157ca52c74871247e3b94f1f5238a05ea408/lib/sqlalchemy/orm/writeonly.py#L93)):

```pytb
sqlalchemy.exc.InvalidRequestError: Attribute User.photos can't load the existing state from the database for this operation; full iteration is not permitted.  If this is a delete operation, configure passive_deletes=True on the User.photos relationship in order to resolve this error.
```

Adding `passive_deletes=True` to the `User.photos` relationship did resolve this error, but why?

```
photos: WriteOnlyMapped['Photo'] = relationship(back_populates='author', cascade='all, delete-orphan', passive_deletes=True)
```

Should the `passive_deletes` always be set to `False` when using the write-only loader? Even if I didn't enable the ON DELETE CASCADE for the foreign key with `on_delete='CASCADE'`?

<details>
<summary>The full trackback</summary>

```pytb

moments/blueprints/user.py:261: in delete_account
    db.session.delete(current_user._get_current_object())
.venv/lib/python3.12/site-packages/sqlalchemy/orm/scoping.py:672: in delete
    return self._proxied.delete(instance)
.venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py:3482: in delete
    self._delete_impl(state, instance, head=True)
.venv/lib/python3.12/site-packages/sqlalchemy/orm/session.py:3509: in _delete_impl
    cascade_states = list(
.venv/lib/python3.12/site-packages/sqlalchemy/orm/mapper.py:3968: in cascade_iterator
    queue = deque(
.venv/lib/python3.12/site-packages/sqlalchemy/orm/relationships.py:1510: in cascade_iterator
    tuples = self._value_as_iterable(
.venv/lib/python3.12/site-packages/sqlalchemy/orm/relationships.py:1486: in _value_as_iterable
    for o in impl.get_collection(state, dict_, x, passive=passive)
.venv/lib/python3.12/site-packages/sqlalchemy/orm/writeonly.py:234: in get_collection
    history = self._get_collection_history(state, passive)
.venv/lib/python3.12/site-packages/sqlalchemy/orm/writeonly.py:404: in _get_collection_history
    return self.collection_history_cls(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <sqlalchemy.orm.writeonly.WriteOnlyHistory object at 0x7fe50552f230>
attr = <sqlalchemy.orm.writeonly.WriteOnlyAttributeImpl object at 0x7fe505e06600>
state = <sqlalchemy.orm.state.InstanceState object at 0x7fe505472f30>, passive = 159
apply_to = <sqlalchemy.orm.writeonly.WriteOnlyHistory object at 0x7fe5052b33e0>

    def __init__(
        self,
        attr: WriteOnlyAttributeImpl,
        state: InstanceState[_T],
        passive: PassiveFlag,
        apply_to: Optional[WriteOnlyHistory[_T]] = None,
    ) -> None:
        if apply_to:
            if passive & PassiveFlag.SQL_OK:
>               raise exc.InvalidRequestError(
                    f"Attribute {attr} can't load the existing state from the "
                    "database for this operation; full iteration is not "
                    "permitted.  If this is a delete operation, configure "
                    f"passive_deletes=True on the {attr} relationship in "
                    "order to resolve this error."
                )
E               sqlalchemy.exc.InvalidRequestError: Attribute User.photos can't load the existing state from the database for this operation; full iteration is not permitted.  If this is a delete operation, configure passive_deletes=True on the User.photos relationship in order to resolve this error.
```
</details>

The same issue also happened for a many-to-many relationship with the association table:

```python
tagging = Table(
    'tagging',
    Base.metadata,
    Column('photo_id', ForeignKey('photo.id'), primary_key=True),
    Column('tag_id', ForeignKey('tag.id'), primary_key=True),
)



class Photo(Base):
    __tablename__ = 'photo'

    id: Mapped[int] = mapped_column(primary_key=True)
    description: Mapped[Optional[str]] = mapped_column(String(500))

    tags: Mapped[list['Tag']] = relationship(secondary=tagging, back_populates='photos')


class Tag(Base):
    __tablename__ = 'tag'

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(64), index=True, unique=True)

    photos: WriteOnlyMapped['Photo'] = relationship(secondary=tagging, back_populates='tags')
```

The records in the association table are managed by SQLAlchemy automatically. When I try to delete a `Tag` object, the related records should also be deleted. If I set the `Tag.photos` relationship's `passive_deletes` to `True`, will it prevent the auto-delete behavior by SQLAlchemy for the association table?

## Additional context

I'm using SQLAlchemy with Flask-SQLAchemy, so the example code has been modified to remove the Flask-SQLAlchemy usage.

Sorry for asking so many questions. I'm writing a book about Python web development, so I hope to address the details of the usage of the write-only loader.

Thanks!

**Comment by zzzeek** [comment, marked as answer]
I'm assuming you meant to say "True".

Yes, setting passive_deletes=True is recommended, this is explained and illustrated in the sole example we have for WriteOnlyMapped in the [documentation](https://docs.sqlalchemy.org/en/20/orm/large_collections.html#write-only-relationships), because otherwise the ORM would like to load in this collection to see what needs to be deleted, which it can't in this case.

> Even if I didn't enable the ON DELETE CASCADE for the foreign key with on_delete='CASCADE'?

that's up to you, if you are separately ensuring that all the related objects are deleted or updated before you delete the parent object, it should be fine

**Reply by greyli** [reply]
> that's up to you, if you are separately ensuring that all the related objects are deleted or updated before you delete the parent object, it should be fine

So for the second example (many-to-many with association table), when using the write-only loader for the relationship in both sides and one side of the object is being deleted, SQLAlchemy won't manage the deletion of the records in the association table automatically like it did for non-write-only loader relationships. I will need to either set the ON DELETE CASCADE for the foreign keys or delete the related records in the association table manually. Am I right?

**Reply by zzzeek** [reply]
I think so yes, or if you had association objects mapped to that table you deleted

**Reply by greyli** [reply]
I see. Thank you very much!
